pkg/volume/metadata/metadata.go

```diff
@@ -0,0 +1,271 @@
+// different (or the file does not exist) return true
+func isFileToGenerate(fileName, values string) bool {
+	if _, err := os.Lstat(fileName); os.IsNotExist(err) {
+		return true
+	}
+	return readFile(fileName) != values
+}
+
+const currentDir = ".current"
+
+// writeDataInTmpDir writes metadata in file
+func (m *metadataVolume) writeDataInTmpDir() (string, error) {
+	errlist := []error{}
+	tempDir, err := ioutil.TempDir(m.GetPath(), "")
+	for fieldReference, fileName := range m.fieldReferenceFileNames {
+		var data string
+		if data, err = fieldpath.ExtractFieldPathAsString(m.pod, fieldReference); err != nil {
+			glog.Error(err)
+			errlist = append(errlist, err)
+		} else {
+			fullPathFile := path.Join(tempDir, fileName)
+			dir, _ := filepath.Split(fullPathFile)
+			if err = os.MkdirAll(dir, os.ModePerm); err != nil {
+				return "", err
+			}
+			if e := ioutil.WriteFile(fullPathFile, []byte(sortLines(data)), 0644); e != nil {
+				glog.Error(e)
+				errlist = append(errlist, err)
+			}
+		}
+	}
+	return tempDir, utilErrors.NewAggregate(errlist)
+}
+
+// updateSymLinks creates the relative links for the all the files
+func (m *metadataVolume) updateSymLinks() error {
+	for _, f := range m.fieldReferenceFileNames {
+		dir, _ := filepath.Split(f)
+		nbOfSubdir := 0
+		if len(dir) > 0 {
+			nbOfSubdir = len(strings.Split(dir, "/")) - 1
+			if e := os.MkdirAll(path.Join(m.GetPath(), dir), os.ModePerm); e != nil {
+				return e
+			}
+		}
+		if _, e := os.Readlink(path.Join(m.GetPath(), f)); e != nil {
+			// link does not exist create it
+			if e := os.Symlink(path.Join(strings.Repeat("../", nbOfSubdir)+currentDir, f), path.Join(m.GetPath(), f)); e != nil {
+				return e
+			}
+		}
+	}
+	return nil
+}
+
+// writeData writes requested metadata in specified files
+func (m *metadataVolume) writeData() error {
+	var tempDir string
+	var err error
+	tempDir, err = m.writeDataInTmpDir()
+	if err != nil {
+		glog.Error(err)
+		return err
+	}
+	_, tempDirBaseName := filepath.Split(tempDir)
+	m.updateSymLinks()
+	var oldTarget string
+	oldTarget, err = os.Readlink(path.Join(m.GetPath(), currentDir))
+	if err == nil {
+		os.Remove(path.Join(m.GetPath(), currentDir)) // remove actual link
+	}
+	err = os.Symlink(tempDirBaseName, path.Join(m.GetPath(), currentDir)) // (re-)link it
+	if err != nil {
+		return err
+	}
+	if len(oldTarget) > 0 {
+		if e := os.RemoveAll(oldTarget); e != nil {
+			return e
+		}
+	}
+	return nil
+}
+
+// isDataChanged iterate over all the entries to check wether at least one
+// file needs to be updated.
+func (m *metadataVolume) isDataChanged() (bool, error) {
+	errlist := []error{}
+	for fieldReference, fileName := range m.fieldReferenceFileNames {
+		var data string
+		var err error
+		if data, err = fieldpath.ExtractFieldPathAsString(m.pod, fieldReference); err != nil {
+			glog.Error(err)
+			errlist = append(errlist, err)
+		}
+		if isFileToGenerate(path.Join(m.GetPath(), fileName), sortLines(data)) {
```

Comment:
What  is sorted here is the content of the fields. The data is emitted in sorted order by `writeDataInTmpDir` (via the same `sortLines` func) and the direct comparison is performed by `isFileToGenerate` func.
Concerning the two pass approach. It's implemented in this way to be sure (aside bug :) ) the content of the tmp directory is consistent.
The idea is:
1) check all the files and verify if at least one changed
2) in case something changed takes another snapshot of requested metadata
When I implemented this I thought about the idea to store the data in a *map[string]string but I don't like the fact that a method called `isDataChanged` or similar returns data to saved later on.